dijet/production/dijet_balance.py

```diff
@@ -31,20 +31,82 @@ def dijet_balance(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
 
     # TODO: for now, this only works for reco level
     jets = events.Jet
+    if self.dataset_inst.is_mc:
+        genJets = events.GenJet
     jets = ak.pad_none(jets, 3)
 
     pt_avg = (events.probe_jet.pt + events.reference_jet.pt) / 2
-    asym = (events.reference_jet.pt - events.probe_jet.pt) / (2 * pt_avg)
+
+    if self.dataset_inst.is_mc:
+        probe_genJet_pt = np.array(
+            [
+                all_genJets[index] for all_genJets, index in zip(genJets.pt, events.probe_jet.genJetIdx)
+            ],
+        )
+        reference_genJet_pt = np.array(
+            [
+                all_genJets[index] for all_genJets, index in zip(genJets.pt, events.reference_jet.genJetIdx)
+            ],
+        )
+        probe_genJet_phi = np.array(
+            [
+                all_genJets[index] for all_genJets, index in zip(genJets.phi, events.probe_jet.genJetIdx)
+            ],
+        )
+        response_probe = ak.where(
+            events.probe_jet.genJetIdx != -1, probe_genJet_pt / events.probe_jet.pt, -100,
+        )
+        response_reference = ak.where(
+            events.reference_jet.genJetIdx != -1, reference_genJet_pt / events.reference_jet.pt, -100,
+        )
+        refAndProbe = (events.reference_jet.genJetIdx != -1) & (events.probe_jet.genJetIdx != -1)
+        pt_avg_gen = ak.where(
+            refAndProbe, np.divide((probe_genJet_pt + reference_genJet_pt), 2), -100,
+        )
+        mpf_gen_val = np.divide(events.GenMET.pt * np.cos(probe_genJet_phi - events.GenMET.phi), (2 * pt_avg_gen))
+        mpf_gen = ak.where(
+            refAndProbe, mpf_gen_val, -100,
+        )
+        mpfx_gen_val = np.divide(events.GenMET.pt * np.sin(probe_genJet_phi - events.GenMET.phi), (2 * pt_avg_gen))
+        mpfx_gen = ak.where(
+            refAndProbe, mpfx_gen_val, -100,
+        )
```

Comment:
This code problematic because it's essentially a Python loop over events. This should be avoided at all costs because it invalidates the benefits of the columnar approach, and with _AwkwardArray_ there are vectorized implementations available that can be used instead of doing `for something in events.column` and are orders of magnitude faster.

Before merging, this would need to be adapted to the proper columnar idioms. I would also keep the convention of using "snake case" variable names for consistency with the rest of the code (i.e. "genJets" -> "gen_jets" etc.). For invalid values one can use the `EMPTY_FLOAT` variable from `columnflow.columnar_util` to be more consistent.

Here's a minimal code suggestion (not tested) that can be used as a starting point:
```suggestion
    if self.dataset_inst.is_mc:

        # check if probe/reference jets have valid gen match
        probe_jet_valid_gen_match = (events.probe_jet.genJetIdx >= 0)
        reference_jet_valid_gen_match = (events.reference_jet.genJetIdx >= 0)
        both_jets_valid_gen_match = (probe_jet_valid_gen_match & reference_jet_valid_gen_match)

        # get gen jet object matched to probe and reference jets
        probe_jet_gen_jet = events.GenJet[events.probe_jet.genJetIdx]
        reference_jet_gen_jet = events.GenJet[events.reference_jet.genJetIdx]

        # calculate reco/gen pt response
        response_probe = ak.where(
            probe_jet_valid_gen_match,
            probe_jet_gen_jet.pt / events.probe_jet.pt,
            EMPTY_FLOAT,
        )
        response_reference = ak.where(
            reference_jet_valid_gen_match,
            reference_jet_gen_jet.pt / events.reference_jet.pt,
            EMPTY_FLOAT,
        )

        # calculate gen-level pT average, MPF, MPFx
        pt_avg_gen = ak.where(
            both_jets_valid_gen_match,
            np.divide(
                probe_jet_gen_jet.pt + reference_jet_gen_jet.pt,
                2,
            ),
            EMPTY_FLOAT,
        )
        mpf_gen = ak.where(
            both_jets_valid_gen_match,
            np.divide(
                events.GenMET.pt * np.cos(probe_jet_gen_jet.phi - events.GenMET.phi),
                2 * pt_avg_gen,
            ),
            EMPTY_FLOAT,
        )
        mpfx_gen = ak.where(
            both_jets_valid_gen_match,
            np.divide(
                events.GenMET.pt * np.sin(probe_jet_gen_jet.phi - events.GenMET.phi),
                2 * pt_avg_gen,
            ),
            EMPTY_FLOAT,
        )
```